Python/pystate.c

```diff
@@ -1195,6 +1195,70 @@ _PyThread_CurrentFrames(void)
     return result;
 }
 
+/* The implementation of sys._current_exceptions(). This is intended to be
+   called with the GIL held, as it will be when called via
+   sys._current_frames(). It's possible it would work fine even without the GIL
+   held, but haven't thought enough about that.
+*/
+PyObject *
+_PyThread_CurrentExceptions(void)
+{
+    PyThreadState *tstate = _PyThreadState_GET();
+    if (_PySys_Audit(tstate, "sys._current_exceptions", NULL) < 0) {
+        return NULL;
+    }
+
+    PyObject *result = PyDict_New();
+    if (result == NULL) {
+        return NULL;
+    }
+
+    /* for i in all interpreters:
+     *     for t in all of i's thread states:
+     *          if t's frame isn't NULL, map t's id to its frame
+     * Because these lists can mutate even when the GIL is held, we
+     * need to grab head_mutex for the duration.
+     */
+    _PyRuntimeState *runtime = tstate->interp->runtime;
+    HEAD_LOCK(runtime);
+    PyInterpreterState *i;
+    for (i = runtime->interpreters.head; i != NULL; i = i->next) {
+        PyThreadState *t;
+        for (t = i->tstate_head; t != NULL; t = t->next) {
+            _PyErr_StackItem *err_info = _PyErr_GetTopmostException(t);
+            if (err_info == NULL) {
+                continue;
+            }
+            PyObject *id = PyLong_FromUnsignedLong(t->thread_id);
```

Comment:
A single thread can be associated to more than one Python thread state. It is possible that two interpreters run in the same thread, just not at the same time.

Reply:
This is also a limitation in `_current_frames` then, right?
I think it'd be fine in this case to mimic what `_current_frames` does in this regard.